File: Reqnroll/Chapter16/16.2-ContextShortcutPattern/Before/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingContext orderingContext, AuthenticationApiDriver authApiDriver,
    OrderingApiDriver orderingApiDriver, KitchenApiDriver kitchenApiDriver)
{
    [Given("the customer has an order that is waiting for pickup")]
    public async Task GivenTheCustomerHasAnOrderThatIsWaitingForPickup()
    {
        await authApiDriver
            .Login(DomainDefaults.CustomerName, DomainDefaults.Password)
            .Execute();
        var orderRequest = new PlaceOrderRequestObjectMother().Build();
        var placedOrder = await orderingApiDriver.PlaceOrder(orderRequest)
            .Execute();
        await authApiDriver
            .Login(DomainDefaults.KitchenStaff, DomainDefaults.Password)
            .Execute();

        Order? takenOrder = null;
        while (takenOrder?.OrderNo != placedOrder.OrderNo)
        {
            takenOrder = await kitchenApiDriver.TakeNextOrder().Execute();
        }

        await kitchenApiDriver.SetReady(placedOrder.OrderNo)
            .Execute();
        orderingContext.CurrentOrderNo = placedOrder.OrderNo;
    }
}

File: CucumberJVM/Chapter16/16.2-ContextShortcutPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.Order;
import com.wimp.app.specs.drivers.*;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.OrderingContext;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final AuthenticationApiDriver authenticationApiDriver;
    private final OrderingApiDriver orderingApiDriver;
    private final KitchenApiDriver kitchenApiDriver;

    public OrderingStepDefinitions(OrderingContext orderingContext, AuthenticationApiDriver authenticationApiDriver, OrderingApiDriver orderingApiDriver, KitchenApiDriver kitchenApiDriver) {
        this.orderingContext = orderingContext;
        this.authenticationApiDriver = authenticationApiDriver;
        this.orderingApiDriver = orderingApiDriver;
        this.kitchenApiDriver = kitchenApiDriver;
    }

    @Given("the customer has an order that is waiting for pickup")
    public void theCustomerHasAnOrderThatIsWaitingForPickup() throws Exception {
        authenticationApiDriver.login(DomainDefaults.CUSTOMER_NAME, DomainDefaults.PASSWORD)
            .execute();
        var orderRequest = new PlaceOrderRequestObjectMother().build();
        var placedOrder = orderingApiDriver.placeOrder(orderRequest).execute();
         authenticationApiDriver.login(DomainDefaults.KITCHEN_STAFF, DomainDefaults.PASSWORD)
            .execute();

        Order takenOrder = null;
        while (takenOrder == null || takenOrder.getOrderNo() != placedOrder.getOrderNo())
        {
            takenOrder = kitchenApiDriver.takeNextOrder().execute();
        }

        kitchenApiDriver.setReady(placedOrder.getOrderNo()).execute();
        orderingContext.setCurrentOrderNo(placedOrder.getOrderNo());
    }
}
